### DISCUSSION TITLE
Generic AST for EcmaScript

### DISCUSSION THREAD

**Discussion by kdy1** [swc-project/swc#7874] [RFC]
I can make EcmaScript AST generic, but it has pros and cons.

It will make, e.g. migrating built tools for Vue or Svelte to Rust much easier, but it requires a decent amount of work, and it will make compilation slower.
So I want to know if there's a team that wants to do that.

# Idea

API will look similar to [`Stylesheet` of `lightningcss`](https://docs.rs/lightningcss/1.0.0-alpha.45/lightningcss/stylesheet/struct.StyleSheet.html), but with much more complexity.

```rust
pub trait AstExt {
    type Stmt;
    type Expr;
    type Pat;
}

pub type Never = Infallible;

pub struct Standard;

impl AstExt for Standard {
    type Stmt = Never;
    type Expr = Never;
    type Pat = Never;
}


pub struct Program<E: AstExt = Standard> {
    Module(Module<E>),
    Script(Script<E>),
}

pub struct Expr<E: AstExt = Standard> {
     // Many generics variants...
    Ext(E::Expr),
}

```

# Pros

## Reduced binary size & Improved performance of core transforms

We can extract TypeScript nodes as extensions, and it will make all other visitors faster and smaller.

# Cons

## Compile time

**Comment by kdy1** [comment]
Note that this will make compilation much slower.

This is cc to hear opinions about the compile time.

@dsherret
@devongovett 
@Boshen

**Reply by devongovett** [reply]
Maybe it's possible to use `dyn`? You lose some performance to dynamic dispatch but would improve compile time. If you expect a lot of different extensions, maybe that's the way to go. For lightningcss we basically only expect 1 per program, so the compile time isn't really affected. It was pretty complicated to pass all the generics everywhere though. 😉

**Reply by kdy1** [reply]
Hmm.. I didn't think deeply about `dyn` because of Wasm plugins, but I found that we don't need them. I'll think about it more. Thank you! I think it will work.

**Reply by kdy1** [reply]
It does not work because of various traits. e.g. PartialEq

**Reply by kdy1** [reply]
Oh I found a way
